File: lightroom_tagger/core/vision_op.py
"""Vision operation engine — single-call provider primitive and persist stage."""

from __future__ import annotations

import inspect
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any, Literal

from lightroom_tagger.core.error_policy import ConsecutiveAbortTracker, ErrorPolicy
from lightroom_tagger.core.fallback import FallbackDispatcher
from lightroom_tagger.core.provider_registry import ProviderRegistry
from lightroom_tagger.core.provider_resolution import Kind, resolve_model

LogCallback = Callable[[str, str], None] | None


@dataclass(frozen=True)
class VisionOpOutcome:
    status: Literal['written', 'skipped', 'failed']
    reason: str | None = None

    @property
    def wrote(self) -> bool:
        return self.status == 'written'


@dataclass
class VisionOpSpec:
    """Single vision-op invocation: resolve → fallback dispatch → parse."""

    resolve_kind: Kind
    operation: str
    provider_id: str | None
    model: str | None
    fn_factory: Callable[[], Callable]
    # Parsers take either ``(raw)`` or ``(raw, provider, model)`` — the extra
    # provider/model are supplied for ops (e.g. scoring) whose parser needs the
    # actual fallback-selected provider to build an LLM repair fixer.
    parse_response: Callable[..., Any]
    log_callback: LogCallback = None
    registry: ProviderRegistry | None = None
    error_policy: ErrorPolicy | None = None
    abort_tracker: ConsecutiveAbortTracker | None = None
    _cleanup: Callable[[], None] | None = field(default=None, repr=False)


def _parser_wants_provider_model(parse_response: Callable[..., Any]) -> bool:
    """True when *parse_response* accepts the ``(raw, provider, model)`` form.

    Determined by the parser's signature (positional params or ``*args``) rather
    than catching ``TypeError``, so a genuine ``TypeError`` raised *inside* the
    parser is never silently swallowed and re-dispatched with the wrong arity.
    """
    try:
        params = inspect.signature(parse_response).parameters.values()
    except (TypeError, ValueError):
        return False
    positional = [
        p for p in params
        if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD)
    ]
    has_var_positional = any(p.kind is p.VAR_POSITIONAL for p in params)
    return has_var_positional or len(positional) >= 3


def run_vision_op(spec: VisionOpSpec) -> tuple[Any, str, str]:
    """Run one vision op: resolve_model → FallbackDispatcher → parse."""
    if spec.registry is None:
        resolved = resolve_model(
            kind=spec.resolve_kind,
            provider_id=spec.provider_id,
            model=spec.model,
        )
        registry = resolved.registry
        provider_id = resolved.provider_id
        model = resolved.model
    else:
        registry = spec.registry
        provider_id = spec.provider_id
        model = spec.model

    dispatcher = FallbackDispatcher(registry, error_policy=spec.error_policy)
    try:
        dispatcher_fn_factory = spec.fn_factory()
        raw, actual_provider, actual_model = dispatcher.call_with_fallback(
            operation=spec.operation,
            fn_factory=dispatcher_fn_factory,
            provider_id=provider_id,
            model=model,
            log_callback=spec.log_callback,
            abort_tracker=spec.abort_tracker,
        )
        if _parser_wants_provider_model(spec.parse_response):
            parsed = spec.parse_response(raw, actual_provider, actual_model)
        else:
            parsed = spec.parse_response(raw)
        return parsed, actual_provider, actual_model
    finally:
        if spec._cleanup is not None:
            spec._cleanup()


def run_vision_op_persist(
    spec: VisionOpSpec,
    *,
    pre_check: Callable[[], VisionOpOutcome | None] | None = None,
    accept_result: Callable[[Any], bool],
    persist: Callable[[Any, str, str], None],
) -> VisionOpOutcome:
    """Pre-checks → core → persist; returns outcome without swallowing exceptions."""
    if pre_check is not None:
        early = pre_check()
        if early is not None:
            return early

    parsed, provider, model = run_vision_op(spec)
    if not accept_result(parsed):
        return VisionOpOutcome(status='skipped', reason='invalid result')
    persist(parsed, provider, model)
    return VisionOpOutcome(status='written')

File: apps/visualizer/backend/src/vision/vision-op.ts
/**
 * The vision operation engine — one provider call plus a persist stage.
 *
 * The shape exists so that "resolve a model, dispatch with fallback, parse the
 * answer" is written once. Describe and score differ only in their prompt, their
 * parser and what they store; every retry, cascade and cancellation rule is
 * shared, which is the only way those rules stay consistent between them.
 */
import { FallbackDispatcher, type FnFactory } from '../providers/fallback.js';
import type { ProviderRegistry } from '../providers/registry.js';
import { resolveModel, type Kind } from '../providers/resolution.js';
import type { CancelCheck, LogCallback } from '../providers/retry.js';

export type VisionOpStatus = 'written' | 'skipped' | 'failed';

export class VisionOpOutcome {
  readonly status: VisionOpStatus;
  readonly reason: string | null;

  constructor(status: VisionOpStatus, reason: string | null = null) {
    this.status = status;
    this.reason = reason;
  }

  get wrote(): boolean {
    return this.status === 'written';
  }
}

export interface VisionOpSpec<TParsed> {
  resolveKind: Kind;
  /** Log label — `"describe"` or `"score"`. */
  operation: string;
  providerId: string | null;
  model: string | null;
  /**
   * Builds the `(client, model) -> call` factory.
   *
   * Called once per op, not once per attempt, so per-op setup — RAW conversion
   * and compression — happens a single time no matter how far the cascade runs.
   */
  fnFactory: () => Promise<FnFactory<string>>;
  /**
   * Parse the raw text.
   *
   * Receives the provider and model that actually served the call, because the
   * scoring parser uses them to build an LLM repair call against the same model.
   */
  parseResponse: (raw: string, provider: string, model: string) => TParsed | Promise<TParsed>;
  logCallback?: LogCallback;
  registry?: ProviderRegistry | null;
  cancelCheck?: CancelCheck;
  /** Always runs, so temp files are removed even when the cascade throws. */
  cleanup?: (() => void | Promise<void>) | null;
}

/** Run one vision op: resolve → dispatch with fallback → parse. */
export async function runVisionOp<TParsed>(
  spec: VisionOpSpec<TParsed>,
): Promise<{ parsed: TParsed; provider: string; model: string }> {
  let registry: ProviderRegistry;
  let providerId: string;
  let model: string;

  if (spec.registry) {
    // A caller that already resolved (the batch handlers do, once per run)
    // passes its registry so this does not re-read providers.json per image.
    registry = spec.registry;
    providerId = spec.providerId ?? '';
    model = spec.model ?? '';
  } else {
    const resolved = await resolveModel({
      kind: spec.resolveKind,
      providerId: spec.providerId,
      model: spec.model,
    });
    ({ registry, providerId, model } = resolved);
  }

  const dispatcher = new FallbackDispatcher(registry);
  try {
    const fnFactory = await spec.fnFactory();
    const { result, providerId: actualProvider, model: actualModel } =
      await dispatcher.callWithFallback({
        operation: spec.operation,
        fnFactory,
        providerId,
        model,
        logCallback: spec.logCallback ?? null,
        cancelCheck: spec.cancelCheck ?? null,
      });
    const parsed = await spec.parseResponse(result, actualProvider, actualModel);
    return { parsed, provider: actualProvider, model: actualModel };
  } finally {
    await spec.cleanup?.();
  }
}

/**
 * Pre-check → run → persist, reporting an outcome without swallowing exceptions.
 *
 * `acceptResult` is the gate that stops an empty or malformed answer being
 * stored: a description row with a blank summary looks identical to a real one in
 * the UI, and would stop the image being retried.
 */
export async function runVisionOpPersist<TParsed>(
  spec: VisionOpSpec<TParsed>,
  hooks: {
    preCheck?: (() => VisionOpOutcome | null) | null;
    acceptResult: (parsed: TParsed) => boolean;
    persist: (parsed: TParsed, provider: string, model: string) => void | Promise<void>;
  },
): Promise<VisionOpOutcome> {
  const early = hooks.preCheck?.();
  if (early) return early;

  const { parsed, provider, model } = await runVisionOp(spec);
  if (!hooks.acceptResult(parsed)) {
    return new VisionOpOutcome('failed', 'invalid result');
  }
  await hooks.persist(parsed, provider, model);
  return new VisionOpOutcome('written');
}
